### ISSUE TITLE
Errors subclassed from ChaliceViewError are not transformed by API Gateway

### ISSUE THREAD

**Issue by nathan-muir** [awslabs/chalice#198]
**Problem**

I want to return errors with a custom "Code", you would expect this would work:

```py
from chalice import Chalice, BadRequestError, ChaliceViewError

class MyCustomError(BadRequestError):
    pass

class AnotherCustomError(ChaliceViewError):
   STATUS_CODE = 418

app = Chalice(app_name='error-test')

@app.route('/example')
def route_example():
  raise MyCustomError("Here's a custom message")

@app.route('/example2')
def route_example2():
  raise AnotherCustomError("I'm a teapot")

```

Running `chalice local`:

```sh
> curl http://localhost:8000/example
{"Code": "MyCustomError", "Message": "MyCustomError: Here's a custom message"}
```

however, once deployed to production:

```sh
> curl http://app-uri/example
{"errorType": "MyCustomError", "errorMessage": "MyCustomError: Here's a custom message", "stackTrace": ...}
```

**Cause**

TL;DR `chalice.app.ALL_ERRORS` != `subclasses(ChaliceViewError)`

In `deployer.py`, `create_error_responses` is only called with the hardcoded list `chalice.app.ALL_ERRORS`.

`create_error_responses` then configures the api-gateway to match only the exact error names are caught and responses are transformed into the `Code` / `Message` format.

However, `chalice.app.Chalice.__call__` passes through any subclass of `ChaliceViewError`, not those just in `chalice.app.ALL_ERRORS`.

Resulting in subclasses of `ChaliceViewError` created in user code not being transformed at all, so they appear like errors when `app.debug = True`

**Work Around**

While the second example doesn't work, you can make the first one work by prefixing the parent error name:

```
class BadRequestErrorMyCustomError(BadRequestError):
    pass
```

This works because the API-Gateway is searching for errors matching  #`BadRequestError.*`

**Possible Solutions**

1. Dis-allow subclassing of `ChaliceViewError` & it's default subclasses
2. Wrap/re-raise any subclass of `ChaliceViewError` that isn't in `ALL_ERRORS`
3. Ensure consistent behaviour between `chalice local` and `chalice deploy` by modelling the same error-matching regexp used by API Gateway Integration Responses.
4. (*Preferred*) Detect subclasses at deploy time, by walking `ChaliceViewError.__subclasses__()`
5. same as  #(4) but using `@chalice.register_error` or similar; has the benefit of allowing errors that don't inherit from `ChaliceViewError`

**Comment by jamesls** [comment]
I like option 4.  This might be a problem if you have custom errors that aren't imported during the deployment process, but that's currently unlikely.  We can always add option (5) if we need more control later.

**Comment by jamesls** [comment]
Just wanted to follow up on this issue.  The recommended way to do this now is using the `Response` class which lets you specify the status code as well as the JSON body:

http://chalice.readthedocs.io/en/latest/quickstart.html#tutorial-customizing-the-http-response
